From b5a66e654e1d423970c8a79d6c892a9778589d5e Mon Sep 17 00:00:00 2001
From: Karthik Kambatla <kasha@apache.org>
Date: Fri, 24 Oct 2014 15:00:12 -0700
Subject: [PATCH 212/596] YARN-2713. "RM Home" link in NM should point to one
 of the RMs in an HA setup. (kasha)

(cherry picked from commit 683897fd028dcc2185383f73b52d15245a69e0cb)
(cherry picked from commit 9af3cfc7eb64389b9e2001f60e7ad810fe260a23)
(cherry picked from commit 589c9f97677f8d22552dec098b58c4f4b2fc7dc0)
---
 .../hadoop/yarn/webapp/util/WebAppUtils.java       |   19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/webapp/util/WebAppUtils.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/webapp/util/WebAppUtils.java
index a8f67ff..3aeb33e 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/webapp/util/WebAppUtils.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/webapp/util/WebAppUtils.java
@@ -143,14 +143,29 @@ public static String getResolvedRMWebAppURLWithoutScheme(Configuration conf) {
   public static String getResolvedRMWebAppURLWithoutScheme(Configuration conf,
       Policy httpPolicy) {
     InetSocketAddress address = null;
+    String rmId = null;
+    if (HAUtil.isHAEnabled(conf)) {
+      // If HA enabled, pick one of the RM-IDs and rely on redirect to go to
+      // the Active RM
+      rmId = (String) HAUtil.getRMHAIds(conf).toArray()[0];
+    }
+
     if (httpPolicy == Policy.HTTPS_ONLY) {
       address =
-          conf.getSocketAddr(YarnConfiguration.RM_WEBAPP_HTTPS_ADDRESS,
+          conf.getSocketAddr(
+              rmId == null
+                  ? YarnConfiguration.RM_WEBAPP_HTTPS_ADDRESS
+                  : HAUtil.addSuffix(
+                  YarnConfiguration.RM_WEBAPP_HTTPS_ADDRESS, rmId),
               YarnConfiguration.DEFAULT_RM_WEBAPP_HTTPS_ADDRESS,
               YarnConfiguration.DEFAULT_RM_WEBAPP_HTTPS_PORT);
     } else {
       address =
-          conf.getSocketAddr(YarnConfiguration.RM_WEBAPP_ADDRESS,
+          conf.getSocketAddr(
+              rmId == null
+                  ? YarnConfiguration.RM_WEBAPP_ADDRESS
+                  : HAUtil.addSuffix(
+                  YarnConfiguration.RM_WEBAPP_ADDRESS, rmId),
               YarnConfiguration.DEFAULT_RM_WEBAPP_ADDRESS,
               YarnConfiguration.DEFAULT_RM_WEBAPP_PORT);      
     }
-- 
1.7.9.5

